src/voter_set.rs

```diff
@@ -12,148 +12,241 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Maintains the `VoterSet` of the blockchain.
-//!
-//! See docs on `VoterSet` for more information.
-
-use crate::std::{self, collections::HashMap, hash::Hash, vec::Vec};
-
-use super::threshold;
-
-/// A voter set, with accompanying indices.
+//! Implementation of a `VoterSet`, representing the complete set
+//! of voters and their weights in the context of a round of the
+//! protocol.
+
+use crate::std::{
+	collections::{HashMap, hash_map::Entry},
+	hash::Hash,
+	num::{NonZeroU64, NonZeroUsize},
+	vec::Vec
+};
+use crate::weights::VoterWeight;
+
+/// A (non-empty) set of voters and associated weights.
+///
+/// A `VoterSet` identifies all voters that are permitted to vote in a round
+/// of the protocol and their associated weights. A `VoterSet` is furthermore
+/// equipped with a total order, given by the ordering of the voter's IDs.
 #[derive(Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "std", derive(Debug))]
 pub struct VoterSet<Id: Hash + Eq> {
-	weights: HashMap<Id, VoterInfo>,
-	voters: Vec<(Id, u64)>,
-	threshold: u64,
+	/// The voters in the set.
+	voters: HashMap<Id, VoterInfo>,
+	/// The total order associated with the keys in `voters`.
+	order: Vec<Id>,
+	/// The required weight threshold for supermajority w.r.t. this set.
+	threshold: VoterWeight,
+	/// The total weight of all voters.
+	total_weight: VoterWeight,
 }
 
 impl<Id: Hash + Eq> VoterSet<Id> {
-	/// Get the voter info for a voter.
-	pub fn info<'a>(&'a self, id: &Id) -> Option<&'a VoterInfo> {
-		self.weights.get(id)
-	}
+	/// Create a voter set from a weight distribution produced by the given iterator.
+	///
+	/// If the distribution contains multiple weights for the same voter ID, they are
+	/// understood to be partial weights and are accumulated. As a result, the
+	/// order in which the iterator produces the weights is irrelevant.
+	///
+	/// Returns `None` if the iterator produced no non-zero weights, i.e.
+	/// the voter set would be empty.
+	///
+	/// # Panics
+	///
+	/// If the total voter weight exceeds `u64::MAX`.
+	///
+	pub fn new<I>(weights: I) -> Option<Self>
+	where
+		Id: Ord + Clone,
+		I: IntoIterator<Item = (Id, u64)>
+	{
+		let weights = weights.into_iter();
+
+		// Populate the voter set, thereby calculating the total weight.
+		let mut voters = HashMap::with_capacity(weights.size_hint().0);
+		let mut total_weight = 0u64;
+		for (id, weight) in weights {
+			if let Some(w) = NonZeroU64::new(weight) {
+				// Prevent construction of inconsistent voter sets by checking
+				// for weight overflow (not just in debug mode). The protocol
+				// should never run with such voter sets.
+				total_weight = total_weight.checked_add(weight).expect("Voter weight overflow");
```

Comment:
the general policy we have for panickers is to avoid unless it's provable they can't trigger. Could we return `None` or an error variant here?

Reply:
Sure, if you prefer that.